app/src/behavior.c

```diff
@@ -39,6 +41,144 @@ const struct device *z_impl_behavior_get_binding(const char *name) {
     return NULL;
 }
 
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+int zmk_behavior_get_empty_param_metadata(const struct device *dev,
+                                          struct behavior_parameter_metadata *metadata) {
+    metadata->sets_len = 0;
+    return 0;
+}
+
+#endif // IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+static int validate_hid_usage(uint16_t usage_page, uint16_t usage_id) {
+    LOG_DBG("Validate usage %d in page %d", usage_id, usage_page);
+    switch (usage_page) {
+    case HID_USAGE_KEY:
+        if (usage_id == 0 || (usage_id > ZMK_HID_KEYBOARD_NKRO_MAX_USAGE &&
+                              usage_id < LEFT_CONTROL && usage_id > RIGHT_GUI)) {
+            return -EINVAL;
+        }
+        break;
+    case HID_USAGE_CONSUMER:
+        if (usage_id >
+            COND_CODE_1(IS_ENABLED(CONFIG_ZMK_HID_CONSUMER_REPORT_USAGES_BASIC), (0xFF), (0xFFF))) {
+            return -EINVAL;
+        }
+        break;
+    default:
+        LOG_WRN("Unsupported HID usage page %d", usage_page);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+int zmk_behavior_validate_param_values(const struct behavior_parameter_value_metadata *values,
```

Comment:
Given the name "validate param values", I first thought this was for validating the `struct behavior_parameter_value_metadata` array itself. Not sure how better to clarify this. Could add some documentation to explain it, or maybe rename it? Some ideas

- `zmk_behavior_matches_param_values()`
- `zmk_behavior_validate_param_with_param_values()`

Reply:
Renamed to `zmk_behavior_check_params_match_metadata` for clarity. That work for you?